### PR TITLE
Enable SSL certifcate verification by default

### PR DESCRIPTION
**Pull request by ibmkendrick**
Change summary:
* Customize StreamsConnection object with requests session arguments
* topology.context.submit config object accept StreamsConnection object
* Update unittests to disable SSL certificate for local streams connection
* Modify unittests in test/python/rest to use env variables to get credentials

Fixes IBMStreams/streamsx.topology#857

### CODE REVIEW

**Comment by ddebrunner** [comment]
Can one of the admins verify this patch?

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py [hunk 1]
```diff
@@ -24,31 +21,39 @@ class StreamsConnection:
...
+        >>> sc = StreamsConnection(username="streamsadmin", password="passw0rd", resource_url=_resource_url, verify=False)
         >>> instances = sc.get_instances()
         >>> jobs_count = 0
         >>> for instance in instances:
         ...    jobs_count += len(instance.get_jobs())
         >>> print("There are " + jobs_count + " jobs across all instances.")
-
     """
-    def __init__(self, username=None, password=None, resource_url=None, config=None, instance_id=None):
+    def __init__(self, username=None, password=None, resource_url=None, config=None, instance_id=None, **session_kwargs):
```
Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: There's no need to call this `something_kwargs` just use a name that describes what its role is.

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py [hunk 2]
```diff
@@ -24,31 +21,39 @@ class StreamsConnection:
...
 
     Example:
+        >>> # If the resource url does not have a valid SSL certificate, you can disable the SSL verification warning
+        >>> from requests.packages import urllib3
+        >>> urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+        >>>
         >>> _resource_url = "https://streamsqse.localdomain:8443/streams/rest/resources"
-        >>> sc = StreamsConnection(username="streamsadmin", password="passw0rd", resource_url=_resource_url)
+        >>> # By default SSL verification is enabled, set verify=False to disable it
+        >>> sc = StreamsConnection(username="streamsadmin", password="passw0rd", resource_url=_resource_url, verify=False)
```
Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: `verify` doesn't seem to exist.

**Comment by ddebrunner** [comment]
This is going to have merge issues with pr #869

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py (hunk 1 above)
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: `session_kwargs` contains the list of keyword arguments that can be supplied to the Requests session object that `StreamsConnection` creates.

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py (hunk 2 above)
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: `verify` is an example of Request session object arguments that can be supplied to `session_kwargs`.

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py (hunk 1 above)
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: An alternative approach would be to define the `StreamsConnection` constructor as:
```
def __init__(self, username=None, password=None, resource_url=None, config=None, instance_id=None, session_args=None)
```
And the usage would then be:
```
StreamsConnection(session_args={'verify': False})
```

Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest.py]: FYI PR #869 is splitting `StreamsConnection` into two objects, `StreamsConnection` and `StreamingAnalyticsConnection`.

**Comment by ddebrunner** [comment]
There are a couple of alternative approaches to passing in a keywords argument that sets arbitrary parameters on the Session object.

1. Have the session object as a public field in `StreamsConnection` allowing arbitrary modification of it.
2. Add an optional `session` parameter to `StreamsConnection`  that is used as the Session object.

Option 1 requires `StreamsConnection` (or its subclasses) not to make any REST access in its constructor, which seems reasonable.

The issue with the keywords approach to me is that it inherently links our choice of parameters with the choice of fields supported by Session, an object we don't control.

**Comment by ibmkendrick** [comment]
Will rework as per suggestion 1:
Have the session object as a public field in `StreamsConnection` allowing arbitrary modification of it.

**Comment by ibmkendrick** [comment]
Ready to be tested and merged.

**Comment by ddebrunner** [comment]
test now

**Comment by ddebrunner** [comment]
Can you remove the `get_session` functions and just use the `session` property.

That's what the proposal was, not a getter method!

https://docs.quantifiedcode.com/python-anti-patterns/correctness/implementing_java-style_getters_and_setters.html

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py [hunk 3]
```diff
@@ -349,7 +352,7 @@ class _DistributedSubmitter(_BaseSubmitter):
     """
     def __init__(self, ctxtype, config, graph, username, password):
         _BaseSubmitter.__init__(self, ctxtype, config, graph)
-        self._streams_connection = None
+        self._streams_connection = config.get(ConfigParams.STREAMS_CONNECTION, None)
```
Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py]: FYI - get without a default uses `None` as the default.

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py [hunk 4]
```diff
@@ -511,6 +514,7 @@ class ConfigParams(object):
     SERVICE_NAME = 'topology.service.name'
     FORCE_REMOTE_BUILD = 'topology.forceRemoteBuild'
     JOB_CONFIG = 'topology.jobConfigOverlays'
+    STREAMS_CONNECTION = 'topology.streamsConnection'
```
Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py]: docstring ?

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/tester.py [hunk 5]
```diff
@@ -97,8 +98,14 @@ def setup_distributed(test):
         test.username = os.getenv("STREAMS_USERNAME", "streamsadmin")
         test.password = os.getenv("STREAMS_PASSWORD", "passw0rd")
 
+        # Disable SSL verification
+        from requests.packages import urllib3
+        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
Comment (ddebrunner) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/tester.py]: What does this do? Is it required, or just nice to have to silence warnings?

Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/tester.py]: I had originally put this in to silence the warning:
```
connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
```

But I don't seem to be getting it anymore even though I remove the code... so I'm going to leave it out for now.

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py (hunk 4 above)
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/context.py]: added

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/tester.py [hunk 6]
```diff
@@ -289,14 +301,15 @@ def _distributed_wait_for_result(self):
         self.result['submission_result'] = self.submission_result
         if self.local_check is not None:
             self._local_thread.join()
-            if self.local_check_exception is not None:
-                raise self.local_check_exception
+        cc._canceljob(self.result)
```
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/tester.py]: The cc._complete is cancelling the job before the `local_check` function completed.  So I have pulled the job cancel logic out and called it after the `local_check` thread has completed.

**Comment by ddebrunner** [comment]
test now

**Comment by ddebrunner** [comment]
test now

**Pull request merged**